omod/src/main/java/org/openmrs/module/datafilter/filter/SubmissionFilter.java

```diff
@@ -0,0 +1,136 @@
+/**
+ * This Source Code Form is subject to the terms of the Mozilla Public License,
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ *
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+package org.openmrs.module.datafilter.filter;
+
+import org.apache.commons.lang.StringUtils;
+import org.openmrs.Location;
+import org.openmrs.OpenmrsObject;
+import org.openmrs.User;
+import org.openmrs.api.LocationService;
+import org.openmrs.api.UserService;
+import org.openmrs.api.context.Context;
+import org.openmrs.module.datafilter.impl.EntityBasisMap;
+import org.openmrs.module.datafilter.impl.api.DataFilterService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.servlet.*;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.util.*;
+import java.util.stream.Collectors;
+
+public class SubmissionFilter implements Filter {
+
+	private static final Logger log = LoggerFactory.getLogger(SubmissionFilter.class);
+
+	private DataFilterService dataFilterService;
+
+	private LocationService locationService;
+
+	private UserService userService;
+
+	public SubmissionFilter() {
+		this.dataFilterService = Context.getRegisteredComponents(DataFilterService.class).get(0);
+		this.locationService = Context.getLocationService();
+		this.userService = Context.getUserService();
+	}
+
+	/**
+	 * @see Filter#init(FilterConfig)
+	 */
+	@Override
+	public void init(FilterConfig filterConfig) {
+		if (log.isDebugEnabled()) {
+			log.debug("Initializing datafilter web filter....");
+		}
+	}
+
+	/**
+	 * @see Filter#doFilter(ServletRequest, ServletResponse, FilterChain)
+	 */
+	@Override
+	public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain)
+			throws IOException, ServletException {
+
+		chain.doFilter(request, response);
+
+		if (((HttpServletRequest) request).getMethod().equals("POST") &&
```

Comment:
All along what I was trying to say was if an exception is thrown anywhere from this line and below or during saving of the mapping, it means the mapping wasn't saved but the other changes made on the form have all already been saved to the DB in OpenMRS core on line 64 above, so you lose the atomicity required within the boundaries of this particular http request.

Reply:
Unless, you create a transactional method and move all the code in the filter inside that method so that all changes pass or fail at once since it would be all within a single outer transaction.

Reply:
Alternatively minus line 64, just wrap all these other lines in a try catch clause, so that in case there is a failure, it has no effect on the rest of the form making it appear to have its own life cycle,I actually like this better because it is a very easy  1 min :) tweak and you can still add an error message in the session object to be displayed in the UI for the user to see that something went wrong and that the rest of the changes were saved.